### PR TITLE
Enable Flash Attention

### PR DESCRIPTION
**Pull request by khatwanimohit**

### CODE REVIEW

**Comment by khatwanimohit** [comment]
Loss curves:

image
image

File: MaxText/layers.py [hunk 1]
```diff
@@ -429,16 +435,49 @@ def swap_dims(x):
...
-        query,
-        key,
-        value,
-        bias=attention_bias,
-        dropout_rng=dropout_rng,
-        dropout_rate=self.dropout_rate,
-        deterministic=deterministic,
-        dtype=self.dtype,
-        float32_logits=self.float32_logits)
+    if cfg.enable_flash_attention:
```
Comment (rwitten) [inline MaxText/layers.py:438]: Can we encapsulate this switch logic into a function?

**Review by rwitten — COMMENTED** [review]
Some nits but the main very serious issue is convergence.

**Comment by rwitten** [comment]
Convergence is so bad that I wonder if we have flash attention being applied significantly incorrectly.

**Comment by khatwanimohit** [comment]
image

**Review by rwitten — APPROVED** [review]
LGTM thanks
